File: web/src/model/victron_controller/dashboard/DischargeTime.ts
import {BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'

export enum DischargeTime {
    At0200 = "At0200",
    At2300 = "At2300"
}

export const DischargeTime_values: ReadonlyArray<DischargeTime> = [
    DischargeTime.At0200,
    DischargeTime.At2300
] as const;

export function DischargeTime_parse(s: string): DischargeTime {
    const found = DischargeTime_values.find(v => v === s);
    if (found === undefined) {
        throw new Error("Unknown DischargeTime variant: " + s);
    }
    return found;
}

export class DischargeTime_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: DischargeTime, writer: BaboonBinWriter): unknown {
        if (this !== DischargeTime_UEBACodec.lazyInstance.value) {
          return DischargeTime_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        switch (value) {
            case "At0200": BinTools.writeByte(writer, 0); break;
                case "At2300": BinTools.writeByte(writer, 1); break;
            default: throw new Error("Unknown enum variant: " + value);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): DischargeTime {
        if (this !== DischargeTime_UEBACodec .lazyInstance.value) {
            return DischargeTime_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const tag = BinTools.readByte(reader);
        switch (tag) {
            case 0: return "At0200" as DischargeTime;
                case 1: return "At2300" as DischargeTime;
            default: throw new Error("Unknown enum variant tag: " + tag);
        }
    }

    public static readonly BaboonDomainVersion = '0.1.0'
    public baboonDomainVersion() {
        return DischargeTime_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return DischargeTime_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#DischargeTime'
    public baboonTypeIdentifier() {
        return DischargeTime_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new DischargeTime_UEBACodec())
    public static get instance(): DischargeTime_UEBACodec {
        return DischargeTime_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/v0_1_0/discharge_time.rs


#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub enum DischargeTime {
    At0200,
    At2300,
}

impl DischargeTime {
    pub fn parse(s: &str) -> Result<Self, String> {
        match s {
            "At0200" => Ok(DischargeTime::At0200),
            "At2300" => Ok(DischargeTime::At2300),
            _ => Err(format!("Unknown variant: {}", s)),
        }
    }

    pub fn all() -> Vec<DischargeTime> {
        vec![
            DischargeTime::At0200,
            DischargeTime::At2300,
        ]
    }
}

impl std::fmt::Display for DischargeTime {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DischargeTime::At0200 => write!(f, "At0200"),
            DischargeTime::At2300 => write!(f, "At2300"),
        }
    }
}

impl std::str::FromStr for DischargeTime {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse(s)
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for DischargeTime {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for DischargeTime {
    fn encode_ueba(&self, _ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            DischargeTime::At0200 => crate::baboon_runtime::bin_tools::write_byte(writer, 0)?,
            DischargeTime::At2300 => crate::baboon_runtime::bin_tools::write_byte(writer, 1)?,
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for DischargeTime {
    fn decode_ueba(_ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let tag = crate::baboon_runtime::bin_tools::read_byte(reader)?;
        match tag {
            0 => Ok(DischargeTime::At0200),
            1 => Ok(DischargeTime::At2300),
            _ => Err(format!("Unknown enum variant tag: {}", tag).into()),
        }
    }
}